squidpy/gr/_ripley.py

```diff
@@ -0,0 +1,245 @@
+        if mode == "F":
+            random = _ppp(hull, 1, n_observations, seed=seed)
+            tree_c = NearestNeighbors(metric=metric, n_neighbors=n_neigh).fit(coord_c)
+            distances, _ = tree_c.kneighbors(random, n_neighbors=n_neigh)
+            bins, obs_stats = _f_g_function(distances.squeeze(), support)
+        elif mode == "G":
+            tree_c = NearestNeighbors(metric=metric, n_neighbors=n_neigh).fit(coord_c)
+            distances, _ = tree_c.kneighbors(coordinates[cluster_idx != i, :], n_neighbors=n_neigh)
+            bins, obs_stats = _f_g_function(distances.squeeze(), support)
+        elif mode == "L":
+            distances = pdist(coord_c, metric=metric)
+            bins, obs_stats = _l_function(distances, support, N, area)
+        obs_arr[i] = obs_stats
+
+    sims = np.empty((n_simulations, len(bins)))
+    pvalues = np.ones((le.classes_.shape[0], len(bins)))
+
+    for i in range(n_simulations):
+        random_i = _ppp(hull, 1, n_observations, seed=seed)
+        if mode == "F":
+            tree_i = NearestNeighbors(metric=metric, n_neighbors=n_neigh).fit(random_i)
+            distances_i, _ = tree_i.kneighbors(random, n_neighbors=1)
+            _, stats_i = _f_g_function(distances_i.squeeze(), support)
+        elif mode == "G":
+            tree_i = NearestNeighbors(metric=metric, n_neighbors=n_neigh).fit(random_i)
+            distances_i, _ = tree_i.kneighbors(coordinates, n_neighbors=1)
+            _, stats_i = _f_g_function(distances_i.squeeze(), support)
+        elif mode == "L":
+            distances_i = pdist(random_i, metric=metric)
+            _, stats_i = _l_function(distances_i, support, N, area)
+
+        for j in range(obs_arr.shape[0]):
+            pvalues[j] += stats_i >= obs_arr[j]
+        sims[i] = stats_i
+
+    pvalues /= n_simulations + 1
+    pvalues = np.minimum(pvalues, 1 - pvalues)
+
+    obs_df = _reshape_res(obs_arr.T, columns=le.classes_, index=bins, var_name=cluster_key)
+    sims_df = _reshape_res(sims.T, columns=np.arange(n_simulations), index=bins, var_name="simulations")
+
+    res = {f"{mode}_stat": obs_df, "sims_stat": sims_df, "bins": bins, "pvalues": pvalues}
+
+    if TYPE_CHECKING:
+        assert isinstance(res, dict)
+
+    if copy:
+        logg.info("Finish", time=start)
+        return res
+
+    _save_data(adata, attr="uns", key=Key.uns.ripley(cluster_key, mode), data=res)
+
+
+def _reshape_res(
+    results: np.ndarray, columns: Union[np.ndarray, List[str]], index: np.ndarray, var_name: str
+) -> pd.DataFrame:
+    df = pd.DataFrame(results, columns=columns, index=index)
+    df.index.set_names(["bins"], inplace=True)
+    df = df.melt(var_name=var_name, value_name="stats", ignore_index=False)
+    df[var_name] = df[var_name].astype("category", copy=True)
+    df.reset_index(inplace=True)
+    return df
+
+
+def _f_g_function(distances: np.ndarray, support: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
```

Comment:
Seems like this could be numba JITted. Maybe leads to perf. gains.

Reply:
experimented a bit an decided to go against it, as there are no clear performance gains
![image](https://user-images.githubusercontent.com/25887487/117979984-c7b27f00-b333-11eb-8071-dd026fe2b319.png)